lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4660,54 +4604,27 @@ HSADispatch::dispatchKernel(hsa_queue_t* lockedHsaQueue, const void *hostKernarg
 
     // Ring door bell
     hsa_signal_store_relaxed(lockedHsaQueue->doorbell_signal, index);
-
-    isDispatched = true;
-
-    return status;
 }
 
 
 
 // wait for the kernel to finish execution
-inline hsa_status_t
-HSADispatch::waitComplete() {
-    _wait_complete_status = HSA_STATUS_SUCCESS;
-    if (!isDispatched)  {
-        _wait_complete_status = HSA_STATUS_ERROR_INVALID_ARGUMENT;
-        return _wait_complete_status;
-    }
-
+inline void
+HSADispatch::wait() {
     if (_signal.handle) {
-        DBOUT(DB_MISC, "wait for kernel dispatch op#" << *this  << " completion with wait flag: " << waitMode << "  signal="<< std::hex  << _signal.handle << std::dec << "\n");
-
         // wait for completion
-        if (hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, uint64_t(-1), waitMode)!=0) {
-            throw Kalmar::runtime_exception("Signal wait returned unexpected value\n", 0);
-        }
-
-        DBOUT (DB_MISC, "complete!\n");
+        hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+        dispose();
```

Comment:
`dispose()` should be called regardless of whether the signal is valid.  Then, the call to `dispose` in `~HSADispatch()` could be removed.

Reply:
If we ever go back to not having every dispatch create a signal, we cannot unconditionally `dispose()` here.  In fact, going back to signal-less dispatches will be an easier change after this PR, so I propose keeping this as-is for now.